enhancements/machine-config/user-data-secret-managed.md

```diff
@@ -0,0 +1,115 @@
+The key point is that the Secret holding the initial stub ignition is completely unmanaged after it's created by the installer.
+In order to upgrade to the new Ignition spec, the MCO needs a way to update the Secret to the new Ignition spec.
+
+
+[1]: https://github.com/openshift/installer/blob/093ca65398fe567bdf63322894496cebbe3d2ade/pkg/asset/ignition/machine/node.go#L30-L36
+[2]: https://github.com/openshift/installer/blob/093ca65398fe567bdf63322894496cebbe3d2ade/pkg/asset/machines/master.go#L161-L170
+[3]: https://github.com/openshift/installer/blob/093ca65398fe567bdf63322894496cebbe3d2ade/pkg/asset/machines/worker.go#L197-L205
+[4]: https://github.com/openshift/installer/blame/093ca65398fe567bdf63322894496cebbe3d2ade/docs/user/aws/install_upi.md#L231-L232
+
+## Motivation
+
+The goal for this enhancement is to have the Secret that contains the stub ignition configuration managed by the MCO. The MCO will later be able to update the stub to a newer version of the spec during an upgrade.
+
+### Goals
+
+MCO will adopt the pointer ignition created by the installer and manage it's lifecycle inside the cluster. The installer is still responsible to create the initial version of the Secret as that is required to support UPI installation by generating the initial ignition configs until the installer provisions the control plane via MAO and not TF. The MCO takes ownership of the Secret containing the stub ignition once the cluster is up and running. When we'll update to the new Ignition version, the MCO can update the Secret.
+
+The final goal for this enhancement is to be able to manage that Secret from the MCO during the ignition migration to the new spec.
+
+If we don't manage that Secret, its content will be on an unsupported version of Ignition that can't provision new machines joining the cluster.
+
+### Non-Goals
+
+To make this simple and to be able to fit this into the release that will bring Ignition v2, we're not considering per-machine Secret(s) as highlighted here https://github.com/openshift/machine-config-operator/issues/683#issuecomment-493193902 and here https://github.com/openshift/machine-config-operator/issues/683#issuecomment-640602493
+
+We don't want to diverge from the current implementation where the Installer still creates the Secret and stub ignition on installation otherwise we'll have to change far more things in order to keep supporting UPI installation that needs ignition configs before starting to install.
+
+## Proposal
+
+The proposal is pretty simple today:
+
+- The installer keeps creating the Secret that holds the stub ignition config (at any given version) under a new name
+- The secret will change name from `<role>-user-data` to `<role>-user-data-managed` which is going to be v3 at a later time before the release
```

Comment:
<role> is insufficient.  We should make a userdata for each machineconfigpool.

Reply:
> is insufficient. We should make a userdata for each machineconfigpool.

and where are we _not_ doing that? I'm not sure what you're referring to, creating userdata for all machineconfigpool in a cluster from the installer or mco isn't something we have now nor we plan to support, we just support master and worker and any custom created (manually...) machinesets will set that to the worker/compute userdata...

Also, if you could expand on your thoughts that would be better next time :)

Reply:
if you're referring to already created custom machinesets (pools are out of the picture here as they have to inherit from the worker MCP) then since we don't support automatic bootimages update, then we don't have this issue as if you're changing the bootimage manually (in 4.6) you have to create this yourself, not the cluster.

Reply:
https://github.com/openshift/enhancements/issues/319

Reply:
> #319

that is not related to this enhancement is it and there's room to improve even when this is in right?? I'll comment on the issue I guess

Reply:
> where are we not doing that since we do this just for master and worker?

This is based on the assumption that there are not custom MachineConfigPools, which is false.

> creating userdata for all machineconfigpool in a cluster from the installer or mco isn't something we have now nor we plan to support

That is insufficient for users that create customer MachineConfigPools.

> that is not related to this enhancement is it and there's room to improve even when this is in right? I'll comment on the issue I guess

Yes and no.  This enhancement is proposing a new naming scheme for userdata secrets that are created either via the MCO or the installer.

I think this enhancement is blocked by the prerequisite of ensuring that any user created MCP and user created MachineSet or Machine that utilizes that MCP is a first-class citizen on the platform.

There be a clear correlation between the user-data and the machine-config-pool for all user data.

Reply:
custom mcps inherit from worker so they _would_ be covered by this no?

Reply:
> custom mcps inherit from worker so they would be covered by this no?

No, they would not, if you want to get the config for the customer MCP on the first boot, eg, custom volumes.

Reply:
I think this is something we *should* fix but given that this proposal would not regress the current state, I wouldn't consider it a blocker.

(I also think most people using custom pools *also* want custom machinesets and for us to automagically tie them together, which is https://github.com/openshift/machine-config-operator/issues/1619 )

Reply:
> I think this enhancement is blocked by the prerequisite of ensuring that any user created MCP and user created MachineSet or Machine that utilizes that MCP is a first-class citizen on the platform.

> There be a clear correlation between the user-data and the machine-config-pool for all user data.

I think we’re in high disagreement then :) how can this be something that blocks that? You’re talking about doing a new thing that was never here and which doesn’t link to this enhancement other than “manage the user data from mco”. What you’re asking here is: while you do this, you might as well do this which is slightly related in the sense that it touches the same subject. That doesn’t at all mean that the enhancement you brought up blocks this. The thing I’m bringing up here is certainly a new feature but nowhere near being something that prevents your enhancement in a later release and definitely higher priority as it touches an area (ignition v3) which is what we want to get in in this release. You’re basically asking to pull in something that doesn’t relate to the ignition migration and also something that we already have a workaround (manual) for.

I certainly agree that improving with your enhancement is something we should do but I keep disagreeing that “while we’re changing the naming here, we should also do that” - these things are internal anyway and we can change that again.
The cost of having a bad migration to v3 because we pull in other bits is definitely not something we’re willing to risk. The migration is complicated as it is as, upgrade side especially, touches things that users touched.

I think @cgwalters replied with basically the same argument I’m bringing here in your enhancement. The mco team hasn’t even the capacity to take on that work along with this.

Reply:
Replying inline also 

> This is based on the assumption that there are not custom MachineConfigPools, which is false.

Mco is still under that assumption as we never worked towards decoupling that yet, we will, your enhancement goes into that direction but we’re not there yet and we’re not holding the migration for that. As Colin states, this isn’t regressing anything, still playing with just masters and workers.

> That is insufficient for users that create customer MachineConfigPools.

This is the case today, these are enhancements so it’s future work to make the situation better but we do have custom mcp today for quite a while so yes it’s a not-so-great support but we have it and this isn’t going towards making that worse but this helps with another thing we want, the new ignition version.

> Yes and no. This enhancement is proposing a new naming scheme for userdata secrets that are created either via the MCO or the installer.

This can be changed again, as this enhancement shows, it impacts nobody to change the naming here, just mco and installer so how does this, again, prevents your enhancement to get in once this gets in.

Reply:
I guess I might just ask then, do we forsee that going with this enhancement will prevent https://github.com/openshift/enhancements/issues/319 from landing at a later time? Enhancements aren’t only about what’s best, sure, if we had capacity we would pull that in as well, but we all got priorities and as long as we don’t prevent further changes to happen I think we can get this before #319

Reply:
All I guess I really care about in this proposal is the naming.  We should have a 1 to 1 mapping of user-data secrets to machine-config-pools.  Since we're touching naming here, we should do that now.

As far as being a prereq, I guess what I'm thinking is, that this enhancement is not much of an enhancement.  The 'real' enhancement is some larger story, and this is an implementation detail (of an intermediate step) rather than a real enhancement.  So sorting these details out is the prereq, not necessarily implementing them.  If we change the names later, more cleanup, more things to migrate.

Not to say that this general problem is unique to this enhancement, but I think it's an increasing trend around this repo that we're treating small stories as an enhancement, so having bigger picture discussions is more difficult.

If there's a compelling reason to not name the user-data the installer creates ```<mcp-name>-user-data``` now, I'm not sure what it is.  "Roles" are more or less going away in k8s as far as I understand them, so mapping user-data to roles long term doesn't make sense.  It does make sense to map the user-data to the config that it's actually requesting IMO.

Reply:
> If there's a compelling reason to not name the user-data the installer creates <mcp-name>-user-data now, I'm not sure what it is. "Roles" are more or less going away in k8s as far as I understand them, so mapping user-data to roles long term doesn't make sense. It does make sense to map the user-data to the config that it's actually requesting IMO.

What I think we’re saying is that we’re not chanting the naming. The role prefix has always been there and we’re not changing them with this, maybe that’s the disconnection we have for this enhancement? What we do with this in 4.6 is appending -managed and that effectively doesn’t change any behavior. It doesn’t make it better to decouple roles as you said but it doesn’t make it worse either so I’m not fully sure again what we’re even discussing about.
Everybody realizes that #319 and what you described in your comment is important but it doesn’t change the fact that this enhancement isn’t about that. This is not changing naming either again, it’s adding a suffix which I can’t see how it would complicate things later on. Having the 1-1 mapping is IMO trivial as well to do in the MCO besides landing this or not.
We’re not postponing a proper enhancement to get away with our stuff, we’re discussing different things, we’re not sticking to the “bad” pattern of not following what an enhancement should be, we’re discussing different things that happen to be related.

Reply:
> Since we're touching naming here, we should do that now.

I’ll keep disagreeing with this also, we’re not changing naming, we’re touching naming in a not-regressive way to implement something else which won’t impact any such change from happening in the near future. What you’re saying is again implying that while we here we might as well do #319 which isn’t, again, the case as these are different thing and touching the naming (in a non regressive way) doesn’t imply that we get to fix #319 which targets another issue - I’ll state this again, this enhancement, be it a real one or just implementation details, is about migrating to ignition v3.

Reply:
fwiw, i agree with @michaelgugino's proposal in #319 . taken by itself the naming change he is proposing sounds like it is not being contested. that said, i think what @runcom is saying here about this enhancement not changing the current behavior with regards to naming makes some sense to me as well from the perspective of planning this work. although i am a little confused about adding `-managed` to some of these names as that sounds like we are starting to get into a naming change.

personally i would love to see #319 be able to be included in this work, but i understand the rationale for separating the two issues.

just my thoughts =)